### PR TITLE
#1785 - admins can add/edit phone numbers for all users

### LINKED ISSUE #1785: admins can add/edit phone numbers for all users
Part of epic #1017

AS an **admin**,
I WANT to add phone numbers to admin, supervisor, and volunteer accounts,
SO THAT they can receive SMS notifications.

### How it works now
Currently, email is the only point of contact associated with user accounts.

### How it should work
There should be a field for mobile phone on the "edit" page for supervisors, volunteers, and admins underneath "Display name". 
Screen Shot 2021-02-25 at 4 04 24 PM (2)

### Acceptance criteria
- [x] a field for mobile numbers is added to the "Edit" page for all user types _except_ for All CASA Admins
- [x] admins can add/edit mobile numbers associated with all users, including themselves
- [x] field should only accept numbers
- [x] field should only accept mobile numbers that are the appropriate length of a US mobile number (ten digits without +1 in the front, 11 digits if 1 is included as the first digit)
- [x] the same mobile number can be shared by different users

### PR DESCRIPTION
**Pull request by xihai01**
### What github issue is this PR for, if any?
Resolves #1785 

### What changed, and why?
Front-end: A phone number form field has been added to the edit profiles of admins, supervisors and volunteers.
Back-end: Phone number column added to users table.

### How will this affect user permissions?
Admins can have the ability to add/edit phone numbers of other admins, supervisors and volunteers.

### How is this tested? (please write tests!) 💖💪
1) Request spec - update routes are tested for all roles to ensure admin can hit these routes to update phone number
2) View spec - check that phone number form fields appear to be editable for admin
3) System spec - check correct error messages show up in UI when phone numbers to be entered are invalid

phone

### Feelings
Me dealing with Git/merging issues (hence I had to close other PR).
fortnight dance

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -12,6 +12,8 @@ class User < ApplicationRecord
 
   validates :email, presence: true
   validates :display_name, presence: true
+
+  validates_with UserValidator
```
Comment (compwron) [inline app/models/user.rb:16]: if we're using the fancy validator, should we put all the existing validators in it also and remove them here?

File: app/policies/user_policy.rb [hunk 2]
```diff
@@ -25,6 +25,14 @@ def update_supervisor_name?
     update_supervisor_email?
   end
 
+  def update_user_setting?
+    if is_supervisor?
+      # allow access to own record or volunteer record
+      return record == user || record.volunteer?
```
Comment (compwron) [inline app/policies/user_policy.rb:31]: permissions issue- must make sure that the volunteer belongs to the correct casa org

File: spec/system/casa_admins/edit_spec.rb [hunk 3]
```diff
@@ -22,13 +24,21 @@
...
     end
   end
 
   context "with invalid data" do
-    it "shows error message for empty email" do
+    let(:role) { "admin" }
+    before do
       visit edit_casa_admin_path(admin)
+      fill_in "Email", with: "newemail@example.com"
+      fill_in "Display name", with: "Kadehara Kazuha"
```
Comment (compwron) [inline spec/system/casa_admins/edit_spec.rb:36]: seems fine https://genshin-impact.fandom.com/wiki/Kaedehara_Kazuha

File: spec/system/supervisors/edit_spec.rb [hunk 4]
```diff
@@ -36,6 +36,16 @@
       expect(page).to have_text("Editing Supervisor")
     end
 
+    context "with invalid data" do
+      let(:role) { "supervisor" }
+      let(:supervisor) { create(:supervisor, display_name: "Lesile Knope", casa_org: organization) }
```
Comment (compwron) [inline spec/system/supervisors/edit_spec.rb:41]: seems fine https://parksandrecreation.fandom.com/wiki/Leslie_Knope

File: spec/system/volunteers/edit_spec.rb [hunk 5]
```diff
@@ -19,18 +19,22 @@
     before do
       sign_in admin
       visit edit_volunteer_path(volunteer)
+      fill_in "volunteer_email", with: "newemail@example.com"
+      fill_in "volunteer_display_name", with: "Kamisato Ayato"
```
Comment (compwron) [inline spec/system/volunteers/edit_spec.rb:23]: seems fine https://genshin-impact.fandom.com/wiki/Kamisato_Ayato

**Review by compwron — CHANGES_REQUESTED** [review]
just need to fix the permissions issue

File: spec/system/casa_admins/edit_spec.rb (hunk 3 above)
Comment (xihai01) [reply spec/system/casa_admins/edit_spec.rb:36]: 🥳 🥳  guess you found my easter eggs

File: app/policies/user_policy.rb (hunk 2 above)
Comment (xihai01) [reply app/policies/user_policy.rb:31]: @compwron I have a question. Where is this volunteer? function defined for "record.volunteer?".

**Review by harsohailB — APPROVED** [review]

File: app/models/user.rb (hunk 1 above)
Comment (xihai01) [reply app/models/user.rb:16]: I think this would be a great issue to create for

File: app/policies/user_policy.rb (hunk 2 above)
Comment (xihai01) [reply app/policies/user_policy.rb:31]: done

**Review by compwron — APPROVED** [review]
This is huge but great!

**Pull request merged**